### ISSUE TITLE
Ability to return current TOTP from secret service api

### ISSUE THREAD

**Issue by olaf-b** [keepassxreboot/keepassxc#7411]
## Summary
Having the keyring report the TOTP can be very useful for automated procedures in a shell. Thus being able to tell KeePassXC to report the TOTP-value for password would make a lot of sense.  Other cases to where it makes sense to report something compound would require that the password return is customizable. So selecting the default {PASSWORD}, or {TOTP}, etc as a configuration option for what a keyring call will return, that would be way cool!

[NOTE]: # ( Why does this feature matter to you? What unique circumstances do you have? )
E.g. automatic connection with vpn using openconnect-sso where there is 2fa required.  I do this in a forked version of openconnect-sso (see my github repo), but I have to send the TOTP secret as password and let the vpn client calculate the TOTP value.  Would be so much better and safer to have the keyring (KeePassXC over libsecret-call) report back the calculated TOTP-value.

**Comment by droidmonkey** [comment]
Actually no, sending a totp value via secret service absolutely invalidates the whole purpose of a second factor. You should use Auto-Type for the totp value to put a human in the loop of that request. We won't be implementing totp over secret service due to this, and I don't think that is even part of the spec.

**Comment by olaf-b** [comment]
@droidmonkey I disagree completely. The point of any *OTP is that the value is not constant. And from the days when the wire was not encrypted, it prevented replay-attacks.  It does not imply a human in the loop. Also, using the auto-type is error-prone and might not work well for console applications.

**Comment by olaf-b** [comment]
So if this is a straight out 'no', and so bluntly rejected,  I will have to reconsider KeePassXC.

**Comment by droidmonkey** [comment]
The challenge is that there is no distinction for TOTP request from secret service. How would that even be handled in this case? Repurposing the password field is not an option. The {TOTP} placeholder is only valid in Auto-Type context. 

I could just as easily leave this open for 4 years and the effect will be the same as an honest no up front.

**Comment by olaf-b** [comment]
Your choice.  I currently use KeePassXC for my passwords, and had to serve the TOTP-secret as a password to let the downstream tools calculate the TOTP-value, i.e. a password-entry in KeePassXC + another password-entry for the TOTP-secret in KeePassXC.  Then libsecret looks up the two different "passwords", and takes one of them for generating the TOTP.  I'd like to be able to have libsecret fetch a "password", but in reality this could be a *OTP or some other compound response composed by {PASSWORD}{ENTER}{TOTP} or similar.  There are many legit use cases for this without telling/restricting people how to deal with their security.  If you find this will never make it into your code, then reject it. If I get the time and inspiration, I will implement in this tool or other.  Pass already has this feature (extension), but doesn't work through libsecret and is also less cross platform.

**Comment by droidmonkey** [comment]
Does your VPN program request a password through libsecret to fill the totp field? What you are asking for is already available through Auto-Type, but targeting console is a bit of a challenge. Albeit much easier with 2.7.0's Auto-Type improvements.

**Comment by olaf-b** [comment]
My workaround is to make secret service provide two passwords: 1) the password, 2) the totp-secret.  Then the application has to calculate the TOTP-value.  What I think would be more secure is to not have the TOTP-secret or password ever leave the keyring/keyvault, and just respond as normal to a password request.  But rather than responding with the {PASSWORD}, it would respond with the {TOTP}.  If the password request response would be configurable just as in the case of AutoType, then this would work well.  In fact, the whole point is to circumvent the need for having any windows in the console case, and also I find that the windows that pop up for e.g. CISCO VPN SAML login will not provide sufficient info to make AutoType work by window identification. Yes, you can do it by manually selecting to AutoFill in the KeePassXC GUI.

**Comment by espenfl** [comment]
I would support adding something like this. Use case would be MFA VPN as well. Considering `pass`, but as @olaf-b says, KeePassXC har some nice benefits of being cross platform.

**Comment by Aetf** [comment]
> But rather than responding with the {PASSWORD}, it would respond with the {TOTP}. If the password request response would be configurable just as in the case of AutoType, then this would work well. 

How would KPXC know which password to respond with? The secret service spec only supports one single session-encrypted password field per entry. So the client has to make two separate requests to get both the password and the totp.

**Comment by michaelk83** [comment]
I think the idea is still to have separate entries for regular password and TOTP (as per [this](https://github.com/keepassxreboot/keepassxc/issues/7411#issuecomment-1039367028) and [this](https://github.com/keepassxreboot/keepassxc/issues/7411#issuecomment-1039539917) comments), but have one of them configured service-side as TOTP. This would move the TOTP derivation service-side, so that the TOTP secret doesn't need to be transferred, only the final time-varying password (this also simplifies the client).

Ideally, the spec would allow the client to configure TOTP entries, but apparently this use case wasn't anticipated.

**Comment by wolandtel** [comment]
I support the request. My VPN uses only TOTP as a password, so I have an entry just for the TOTP.

**Comment by Joshix-1** [comment]
> The {TOTP} placeholder is only valid in Auto-Type context.

That's not correct. It's also valid in the password field. When accessing the entry using the secret service it gets correctly replaced. That has been a feature since secret service support has been added in https://github.com/keepassxreboot/keepassxc/commit/e121f4bc2883fda73a495546ea8b8833e51ca8ae#diff-1cb9c38e9eaf2c9aca0e263ad6d860ce11dccd4d4ed344b36da4bdeb251b95afR415

It's just a bit annoying that the placeholder gets easily overridden if the program writes the token back. Would be cool if there was an option to prevent overriding placeholders.

**Comment by droidmonkey** [comment]
That comment is pretty old so not sure if that changed in between

**Comment by Joshix-1** [comment]
The commit I linked (which added the feature) is from 2019, 3 years older than your comment.

**Comment by droidmonkey** [comment]
Ok,  the fact still stands it's not a feature of the secret service api
